markets/dagstore/piecereader.go

```diff
@@ -0,0 +1,142 @@
+package dagstore
+
+import (
+	"context"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/dagstore/mount"
+	"github.com/filecoin-project/go-state-types/abi"
+)
+
+// For small read skips, it's faster to "burn" some bytes than to setup new sector reader.
+// Assuming 1ms stream seek latency, and 1G/s stream rate, we're willing to discard up to 1 MiB.
+var MaxPieceReaderBurnBytes int64 = 1 << 20 // 1M
+
+type pieceReader struct {
+	ctx      context.Context
+	api      MinerAPI
+	pieceCid cid.Cid
+	len      abi.UnpaddedPieceSize
+
+	closed bool
+	seqAt  int64 // next byte to be read by io.Reader
+
+	r   io.ReadCloser
+	rAt int64
+}
+
+func (p *pieceReader) init() (_ *pieceReader, err error) {
+	p.rAt = 0
+	p.r, p.len, err = p.api.FetchUnsealedPiece(p.ctx, p.pieceCid, uint64(p.rAt))
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+func (p *pieceReader) check() error {
+	if p.closed {
+		return xerrors.Errorf("reader closed")
+	}
+
+	return nil
+}
+
+func (p *pieceReader) Close() error {
+	if err := p.check(); err != nil {
+		return err
+	}
+
+	if p.r != nil {
+		if err := p.r.Close(); err != nil {
+			return err
+		}
+		p.r = nil
+	}
+
+	return nil
+}
+
+func (p *pieceReader) Read(b []byte) (int, error) {
+	if err := p.check(); err != nil {
+		return 0, err
+	}
+
+	n, err := p.ReadAt(b, p.seqAt)
+	p.seqAt += int64(n)
+	return n, err
+}
+
+func (p *pieceReader) Seek(offset int64, whence int) (int64, error) {
+	if err := p.check(); err != nil {
+		return 0, err
+	}
+
+	switch whence {
+	case io.SeekStart:
+		p.seqAt = offset
+	case io.SeekCurrent:
+		p.seqAt += offset
+	case io.SeekEnd:
+		p.seqAt = int64(p.len) + offset
+	default:
+		return 0, xerrors.Errorf("bad whence")
+	}
+
+	return p.seqAt, nil
+}
+
+func (p *pieceReader) ReadAt(b []byte, off int64) (n int, err error) {
+	if err := p.check(); err != nil {
+		return 0, err
+	}
+
+	// get the backing reader into the correct position
+	if p.r == nil {
+		p.rAt = MaxPieceReaderBurnBytes * -2
+	}
```

Comment:
I'm not sure I understand what the purpose of this clause is; It seems that if `p.r` is `nil`, `p.rAt` will be set to `MaxPieceReaderBurnBytes * -2` and so the `if` statement on line 105 will always be true.
If so, why not just include `p.r == nil` in the `if` statement on line 105?

Reply:
Yeah, no good reason for this to be like that